helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,332 @@
+        focused: bool,
+        spinners: &'a ProgressSpinners,
+    ) -> Self {
+        RenderContext {
+            doc,
+            view,
+            theme,
+            focused,
+            spinners,
+            parts: RenderBuffer::default(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RenderBuffer<'a> {
+    pub left: Spans<'a>,
+    pub center: Spans<'a>,
+    pub right: Spans<'a>,
+}
+
+pub struct StatusLine;
+
+impl StatusLine {
+    pub fn render(
+        editor: &Editor,
+        context: &mut RenderContext,
+        viewport: Rect,
+        surface: &mut Surface,
+    ) {
+        let base_style = if context.focused {
+            context.theme.get("ui.statusline")
+        } else {
+            context.theme.get("ui.statusline.inactive")
+        };
+
+        surface.set_style(viewport.with_height(1), base_style);
+
+        let write_left = |context: &mut RenderContext, text, style| {
+            Self::append(&mut context.parts.left, text, &base_style, style)
+        };
+        let write_center = |context: &mut RenderContext, text, style| {
+            Self::append(&mut context.parts.center, text, &base_style, style)
+        };
+        let write_right = |context: &mut RenderContext, text, style| {
+            Self::append(&mut context.parts.right, text, &base_style, style)
+        };
+
+        // Left side of the status line.
+
+        let element_ids = &editor.config().status_line.left;
+        element_ids
+            .iter()
+            .map(|element_id| Self::get_render_function(*element_id))
+            .for_each(|render| render(context, write_left));
+
+        surface.set_spans(
+            viewport.x,
+            viewport.y,
+            &context.parts.left,
+            context.parts.left.width() as u16,
+        );
+
+        // Right side of the status line.
+
+        let element_ids = &editor.config().status_line.right;
+        element_ids
+            .iter()
+            .map(|element_id| Self::get_render_function(*element_id))
+            .for_each(|render| render(context, write_right));
+
+        surface.set_spans(
+            viewport.x
+                + viewport
+                    .width
+                    .saturating_sub(context.parts.right.width() as u16),
+            viewport.y,
+            &context.parts.right,
+            context.parts.right.width() as u16,
+        );
+
+        // Center of the status line.
+
+        let element_ids = &editor.config().status_line.center;
+        element_ids
+            .iter()
+            .map(|element_id| Self::get_render_function(*element_id))
+            .for_each(|render| render(context, write_center));
+
+        // Width of the empty space between the left and center area and between the center and right area.
+        let spacing = 1u16;
+
+        let edge_width = context.parts.left.width().max(context.parts.right.width()) as u16;
+        let center_max_width = viewport.width - (2 * edge_width + 2 * spacing);
```

Comment:
With a small enough window (or a large sections) this overflows for me, resulting in a panic. I'm not sure how we want to prioritize the different sections when there isn't enough space for everything. I'm using `saturating_sub` as a quick fix in my own branch, but that doesn't look particularly good when the left section is larger than the right, as space gets wasted between the center and the right...

Reply:
Thanks for the heads up! I'll look at this. A panic due to a certain configuration or due to a too small window must not happen.

I'm afraid we won't get around truncating elements, and I think the priority would be left > right > center (somehow feels more intuitive).

I'd also consider differentiating between a shortened version of elements (e.g. just a file name instead of a file path) when we detect there isn't enough space. But this would probably better fit in another PR.

Reply:
Alright, I added a `saturating_sub()` too: I can definitely see why the subtraction would cause an overflow.

Moreover, it's entirely correct for `center_max_width` to be `0` if there's not enough space left.

Changed in b2cc7f5.